src/cpp/core/r_util/RActiveSessionsStorage.cpp

```diff
@@ -0,0 +1,114 @@
+/*
+ * RActiveSessionsStorage.cpp
+ *
+ * Copyright (C) 2022 by RStudio, PBC
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+#include <core/r_util/RActiveSessionsStorage.hpp>
+
+#include <core/r_util/RActiveSessionStorage.hpp>
+#include <core/r_util/RActiveSessions.hpp>
+
+#include <algorithm>
+
+namespace rstudio {
+namespace core {
+namespace r_util {
+
+namespace
+{
+
+FilePath getSessionDirPath(const FilePath& storagePath, const std::string& sessionId)
+{
+   return storagePath.completeChildPath(kSessionDirPrefix + sessionId);
+}
+
+} // anonymous namespace
+
+FileActiveSessionsStorage::FileActiveSessionsStorage(const FilePath& rootStoragePath)
+{
+   storagePath_ = ActiveSessions::storagePath(rootStoragePath);
+   Error error = storagePath_.ensureDirectory();
+   if (error)
+   {
+      LOG_ERROR(error);
+   }
+}
+
+bool FileActiveSessionsStorage::hasSessionId(const std::string& sessionId) const
+{
+   FilePath dir = getSessionDirPath(storagePath_, sessionId);
+   return dir.exists();
+}
+
+core::Error FileActiveSessionsStorage::createSession(const std::string& id, std::map<std::string, std::string> initialProperties)
+{
+   FilePath sessionScratchPath = storagePath_.completeChildPath(kSessionDirPrefix + id);
+   Error error = sessionScratchPath.ensureDirectory();
+
+   if(error)
+   {
+      return error;
+   }
+
+   FileActiveSessionStorage session = FileActiveSessionStorage{sessionScratchPath};
+   error = session.writeProperties(initialProperties);
+   return error;
```

Comment:
Nit: fewer lines, possibly one less copy (although the compiler probably already removed it)
```suggestion
   return session.writeProperties(initialProperties);
```